pkg/storaged/pages.jsx

```diff
@@ -0,0 +1,691 @@
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from "cockpit";
+import React, { useState, useRef } from "react";
+import client from "./client";
+import { useEvent } from "hooks.js";
+
+import { AlertGroup } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import { Card, CardHeader, CardTitle, CardBody } from "@patternfly/react-core/dist/esm/components/Card/index.js";
+import { DropdownGroup, DropdownList } from '@patternfly/react-core/dist/esm/components/Dropdown/index.js';
+import { Stack, StackItem } from "@patternfly/react-core/dist/esm/layouts/Stack/index.js";
+import { Split, SplitItem } from "@patternfly/react-core/dist/esm/layouts/Split/index.js";
+import { Bullseye } from "@patternfly/react-core/dist/esm/layouts/Bullseye/index.js";
+import { Button } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import { Table, Thead, Tbody, Tr, Th, Td } from '@patternfly/react-table';
+import { EmptyState, EmptyStateBody } from "@patternfly/react-core/dist/esm/components/EmptyState/index.js";
+import { ExclamationTriangleIcon, ExclamationCircleIcon } from "@patternfly/react-icons";
+import { Page, PageBreadcrumb, PageSection } from "@patternfly/react-core/dist/esm/components/Page/index.js";
+import { Breadcrumb, BreadcrumbItem } from "@patternfly/react-core/dist/esm/components/Breadcrumb/index.js";
+import { Spinner } from "@patternfly/react-core/dist/esm/components/Spinner/index.js";
+import { DescriptionListDescription, DescriptionListGroup, DescriptionListTerm } from "@patternfly/react-core/dist/esm/components/DescriptionList/index.js";
+import { Flex, FlexItem } from "@patternfly/react-core/dist/esm/layouts/Flex/index.js";
+
+import { decode_filename, block_short_name, fmt_size } from "./utils.js";
+import { StorageButton, StorageBarMenu, StorageMenuItem, StorageSize } from "./storage-controls.jsx";
+import { MultipathAlert } from "./multipath.jsx";
+import { JobsPanel } from "./jobs-panel.jsx";
+
+const _ = cockpit.gettext;
+
+let pages = null;
+let crossrefs = null;
+
+export function reset_pages() {
+    pages = new Map();
+    crossrefs = new Map();
+}
+
+function name_from_card(card) {
+    if (!card)
+        return null;
+    return name_from_card(card.next) || card.page_name;
+}
+
+function icon_from_card(card) {
+    if (!card)
+        return null;
+    return icon_from_card(card.next) || card.page_icon;
+}
+
+function key_from_card(card) {
+    if (!card)
+        return null;
+    return key_from_card(card.next) || card.page_key;
+}
+
+function location_from_card(card) {
+    if (!card)
+        return null;
+    return location_from_card(card.next) || card.page_location;
+}
+
+function size_from_card(card) {
+    if (!card)
+        return null;
+    if (card.page_size)
+        return card.page_size;
+    return size_from_card(card.next);
+}
+
+export function new_page(parent, card, options) {
```

Comment:
The new_{page,card} functions could indeed benefit from some high-level comment about how the whole structure is being put together.  For example, lvm2/unsupported-logical-volume.jsx `make_unsupported_logical_volume_page()` calls `new_page()` but entirely ignores its return value, and does not return anything itself either. At first this looks quite magic, until I realized that it inserts itself into `parent`, i.e. finds its little place in the storage forest.

This is almost a new UI framework, and a bit unusual to follow what actually happens.